### ISSUE TITLE
Strange 3D surface from body normal maps

### ISSUE THREAD

**Issue by YuliangXiu** [hoshino042/NormalIntegration#3]
Firstly, thanks for your code, I am applying your normal integration algorithm to my task, *refine coarse geometry with fine normal maps*

I run this script on body and cloth normal maps and reconstructed the surfaces as follows

```bash
python methods/orthographic_five_point_plane_fitting.py --path slack_trial2-000150.npy
```

|normal map|surface-view1|surface-view2|
|:--:|:--:|:--:|
|![slack_trial2-000150_normal_F](https://user-images.githubusercontent.com/7944350/179257444-2367b95d-361b-4c99-b47d-e25d2621060f.png)|![snapshot02_L01](https://user-images.githubusercontent.com/7944350/179258419-171bbad5-e96d-4a58-baa1-2f79ae2be306.png)|![snapshot03_L01](https://user-images.githubusercontent.com/7944350/179258447-dd0936ad-73eb-4749-96ec-4eaeb5873a6e.png)|
|![slack_trial2-000150_T_normal_F](https://user-images.githubusercontent.com/7944350/179257779-8f915e0c-f680-4dd5-966b-f0994217697e.png)|![snapshot02_L00](https://user-images.githubusercontent.com/7944350/179258149-ff998ba3-dd3b-4237-ba98-1e7e22a3c182.png)|![snapshot03_L00](https://user-images.githubusercontent.com/7944350/179258244-2f19f9bc-061a-409f-916d-bdf7a3444272.png)|

It seems that the local offsets are correct, but the global translation is totally wrong. If I already have the **coarse geometry+fine normal maps** of the human, what if I want to introduce the coarse geometry into the optimization process? Should I replace the $d$ with point-to-origin distance $d_o$ computed from coarse geometry?

![image](https://user-images.githubusercontent.com/7944350/179261226-0cfd188b-18d9-4076-98a1-13c8bb4fe546.png)

[human.zip](https://github.com/hoshino042/NormalIntegration/files/9122404/human.zip)

npy file dict keys: 
- clothed human: "normal_map_F", "normal_map_B", "mask"
- coarse body: "T_normal_F", "T_normal_B", "T_mask"

------------------------
UPDATE:

OMG, I tried your ECCV 2022 work, and the results look much much much better
![snapshot04](https://user-images.githubusercontent.com/7944350/179267514-9b7f9783-e369-4e03-8a38-ba088c3f9054.png)

I cannot wait to ask what's the main difference between the *Inverse Plane Fitting* and *Bilateral Normal Integration*?

**Comment by xucao-42** [comment]
Hi, 
I think the distortion stems from the inconsistent coordinate system.  I just pushed a new commit to resolve this issue. Or you can use the methods in this repo in the same way as bilateral normal integration: passing the folder path containing the normal map and the mask to the script. It should work well, as shown below.
CVPR 2021 result:
![est_surface_orthographic_five_point_plane_fitting](https://user-images.githubusercontent.com/28819074/179341399-9e9d84fa-2670-445f-983f-2a399383792a.png)
ECCV 2022 result:
![mesh_k_2](https://user-images.githubusercontent.com/28819074/179341402-a2bb3341-d2dd-4ec7-a64d-91318aac02fa.png)

The major difference between the two works is whether possible discontinuities (i.e. depth gap/edges) can be preserved.
CVPR 2021 method can only recover a smooth (continuous everywhere) surface; ECCV 2022 method can further recover large depth gaps if it exists, as shown below.
CVPR 2021:
![est_surface_orthographic_five_point_plane_fitting](https://user-images.githubusercontent.com/28819074/179341515-644f0216-ce5e-4fd5-b27c-a8b8af923c4e.png)
ECCV 2022:
![mesh_k_2](https://user-images.githubusercontent.com/28819074/179341521-22595381-015a-4ef4-bdeb-4d787187ecd8.png)

CVPR 2021: 
![est_surface_orthographic_five_point_plane_fitting](https://user-images.githubusercontent.com/28819074/179341553-1ba74d98-8a6c-4218-bf32-cd592659a1ff.png)
ECCV 2022: 
![mesh_k_2](https://user-images.githubusercontent.com/28819074/179341560-cf3edc2e-d31d-4924-b922-688d465d1a52.png)

However, discontinuity preservation comes with a time cost that, CVPR 2021 method solves a linear system once, while ECCV 2022 method iteratively solves the linear system for dozens of steps. But if the iteration number is set as 1, the two methods should give almost the same result: the reconstructed surface is continuous everywhere. So we believe that given the ECCV 2022 work, there is no reason to use this repo anymore in practice.

Regarding refining coarse geometry, I will soon update the ECCV 2022 repo to include that feature.
The idea is the same as [this paper](https://w3.impa.br/~diego/publications/NehEtAl05.pdf): Using depth map as a regularization term.

Hope it helps.

**Comment by YuliangXiu** [comment]
Really appreciate your insightful and detailed response. I think your ECCV 2022 work really deals with my problem.

Looking forward to the *refining coarse geometry* module, this will help a lot on my project.

Here is a very simple demo for ICON+BNI (Bilateral Normal Integration)
- white: ICON reconstruction, relatively coarse
- green: front surface optimized from front normal prediction using BNI, much more detailed wrinkles
- orange: back surface optimized from back normal prediction using BNI, much more detailed wrinkles

https://user-images.githubusercontent.com/7944350/179357926-1acfda47-66d1-4b2d-b62a-d443ed833e25.mp4

I also notice the *single hyperparameter* claim in your ECCV 2022, does $k$ control the discontinuity depth?

https://user-images.githubusercontent.com/7944350/179358030-3db15767-cedd-4fa8-910e-8fd1fec20587.mp4

**Comment by xucao-42** [comment]
This demo looks really cool. I recently worked on multiview normal map fusion, but I think it's difficult to make it a standalone paper because, basically it's like just an application of bilateral normal integration. If you would like to, maybe we can collaborate to make your idea work. 

Yes, the hyperparameter k comes from the sigmoid function and controls how easily the discontinuity can be recovered. The larger k is, the easier depth gap is to be recovered. In general, we found that k=2 works in most cases.  And k is fixed during the iteration. The  depth gap is gradually recovered because BNI is solving a weighted least-squares problem at each iteration, and the weights are updated at each iteration using the currently estimated depth map (https://github.com/hoshino042/bilateral_normal_integration/blob/main/bilateral_normal_integration.py#L204-L208).
